core/src/cost_tracker.rs

```diff
@@ -28,6 +43,28 @@ impl CostTracker {
         }
     }
 
+    pub fn would_transaction_fit(&self, transaction: &Transaction) -> Result<(), &'static str> {
+        let mut cost_model = self.cost_model.write().unwrap();
+        let tx_cost = cost_model.calculate_cost(transaction);
```

Comment:
Seems like we don't need to hold the lock for the duration of this method?
```rust
let tx_cost = self
    .cost_model
    .write()
    .unwrap()
    .calculate_cost(transaction);
```

`add_transaction_cost()` below looks like it can take a similar treatment

Reply:
Right, this is a bit tricky because `cost_model` has this signature for calculate_cost: 
```
    pub fn calculate_cost(&mut self, transaction: &Transaction) -> &TransactionCost {
```
It reuses a preallocated `TransactionCost`, and returns calculation result as a ref to it. The motivation is `calculate_cost` is called for every transaction, therefore it is in hot path. It saves noticeable overhead by avoiding allocating `TransactionCost` at heap for each call. But this requires the lock to live as long as `tx_cost`. Otherwise Rust complains: 
```
error[E0716]: temporary value dropped while borrowed
  --> core/src/cost_tracker.rs:47:23
   |
47 |         let tx_cost = self.cost_model.write().unwrap().calculate_cost(transaction);
   |                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^                            - temporary value is freed at the end of this statement
   |                       |
   |                       creates a temporary which is freed while still in use
48 |         self.would_fit(
49 |             &tx_cost.writable_accounts,
   |             -------------------------- borrow later used here
   |
   = note: consider using a `let` binding to create a longer lived value
```
I compared the execution timing between two approaches, reuse heap object and holding lock a bit longer had much better performance. 
I am glad you pointed it out, maybe I over complicated it, and there is a better way to avoid many head alloc.